corehq/apps/export/multithreaded.py

```diff
@@ -0,0 +1,379 @@
+"""
+This package contains classes and function for processing ExportInstances using
+multiple threads.
+
+To rebuild an export run the following:
+
+    rebuild_export_mutithreaded(export_instance_id, num_threads, page_size)
+
+You can also use the MultithreadedExporter class to have more control over the process.
+See the 'process_skipped_pages' management command for an example.
+
+The export works as follows:
+  * Dump raw docs from ES into files of size N docs
+  * Once each file is complete add it to a multiprocessing Queue
+  * Pool of X processes listen to queue and process the dump file
+  * Results returned back to the main process
+    * Unsuccessful results can be retried
+  * Add successful pages to final ZIP archive
+  * Add raw data dumps for unsuccessful pages to final ZIP archive
+"""
+import gzip
+import json
+import logging
+import multiprocessing
+import os
+import tempfile
+import time
+import zipfile
+from Queue import Empty
+from collections import namedtuple
+from datetime import timedelta
+
+from corehq.apps.export.dbaccessors import get_properly_wrapped_export_instance
+from corehq.apps.export.export import (
+    ExportFile, _get_writer, _save_export_payload, get_export_size, _get_export_documents)
+from corehq.apps.export.export import _write_export_instance
+from corehq.elastic import ScanResult
+from corehq.util.files import safe_filename
+from couchexport.export import get_writer
+from couchexport.writers import ZippedExportWriter
+
+UNPROCESSED_PAGES_DIR = 'unprocessed'
+
+logger = logging.getLogger(__name__)
+
+
+ProgressValue = namedtuple('ProgressValue', 'page progress total')
+
+
+class BaseResult(object):
+    success = False
+
+    def __init__(self, page_number, page_path, page_size):
+        self.page = page_number
+        self.path = page_path
+        self.page_size = page_size
+
+
+class SuccessResult(BaseResult):
+    success = True
+
+
+class RetryResult(BaseResult):
+    def __init__(self, page_number, page_path, page_size, retry_count):
+        super(RetryResult, self).__init__(page_number, page_path, page_size)
+        self.retry_count = retry_count
+
+
+class QueuedResult(RetryResult):
+    def __init__(self, async_result, page_number, page_path, page_size, retry_count):
+        super(QueuedResult, self).__init__(page_number, page_path, page_size, retry_count)
+        self.async_result = async_result
+
+
+class OutputPaginator(object):
+    """Helper class to paginate raw export output"""
+    def __init__(self, export_id):
+        self.export_id = export_id
+        self.page = 0
+        self.page_size = 0
+        self.file = None
+
+    def __enter__(self):
+        self._new_file()
+
+    def _new_file(self):
+        if self.file:
+            self.file.close()
+        self.path = tempfile.mktemp()
+        self.file = gzip.open(self.path, 'wb')
```

Comment:
`tempfile.mktemp` is deprecated. Better to use `NamedTemporaryFile`:
```py
fileobj = tempfile.NamedTemporaryFile(mode='wb', delete=False)
self.path = fileobj.name
self.file = gzip.GzipFile(fileobj=fileobj)
```

Reply:
Noted thanks. I had changed other places to use `mkstemp`.